### PR TITLE
Support builds via depot.dev

### PR DESCRIPTION
**Pull request by jsierles**
This borrows code from dockerfileBuilder, but has some differences as depot is remote and can build and push to the fly registry directly.

The CLI arguments followed the style of the nixpacks flags.

To deploy using depot:
- flyctl deploy --depot

To run a machine:
* flyctl machines run --build-depot

### CODE REVIEW

**Comment by billyb2** [comment]
related to:
https://github.com/superfly/fly-go/pull/86
https://github.com/superfly/web/pull/3297/

**Comment by KTanAug21** [comment]
Hello! 

I'm trying to review this PR by checking if it works locally:
✅ 1. Check if `go build` works 
-works great with ` go build -o bin/flyctl.exe .`

❌ 2. Use the built flyctl.exe to run `flyctl.exe deploy --depot` in a local Laravel Fly App directory:
       - running the command gives me an error:
```
/.../newapp$ /.../go/flyctl/bin/flyctl.exe  deploy --depot
==> Verifying app config
Validating /home/admin_kath/development/projects/php/tests/newapp/fly.toml
✓ Configuration is valid
--> Verified app config
==> Building image
==> Building image
Error: failed to fetch an image or build from source: error building: returned error 500 Internal Server Error: {"errors":[{"message":"You hit a Fly API error with request ID: 01J3ZWBYA2E6V29C7SB41NQ448-ams","extensions":{"code":"SERVER_ERROR","fly_request_id":"01J3ZWBYA2E6V29C7SB41NQ448-ams"}}],"data":{}}
```
       - just a note: running `fly deploy` in that local Laravel Fly App directory succeeds though
       - what can I do to make this command run successfully? Did I miss anything?

**Comment by billyb2** [comment]
> Hello!
> 
> I'm trying to review this PR by checking if it works locally: ✅ 1. Check if `go build` works -works great with ` go build -o bin/flyctl.exe .`
> 
> ❌ 2. Use the built flyctl.exe to run `flyctl.exe deploy --depot` in a local Laravel Fly App directory: - running the command gives me an error:
> 
> ```
> /.../newapp$ /.../go/flyctl/bin/flyctl.exe  deploy --depot
> ==> Verifying app config
> Validating /home/admin_kath/development/projects/php/tests/newapp/fly.toml
> ✓ Configuration is valid
> --> Verified app config
> ==> Building image
> ==> Building image
> Error: failed to fetch an image or build from source: error building: returned error 500 Internal Server Error: {"errors":[{"message":"You hit a Fly API error with request ID: 01J3ZWBYA2E6V29C7SB41NQ448-ams","extensions":{"code":"SERVER_ERROR","fly_request_id":"01J3ZWBYA2E6V29C7SB41NQ448-ams"}}],"data":{}}
> ```
> 
> ```
>    - just a note: running `fly deploy` in that local Laravel Fly App directory succeeds though
>    - what can I do to make this command run successfully? Did I miss anything?
> ```

ah whoops. a secret accidentally got overridden with the incorrect value. just fixed that

**Comment by KTanAug21** [comment]
Awesome! I've re-run and the deploy succeeded with depot! 🎉

**Comment by KTanAug21** [comment]
Here are the manual tests I've done to verify the commands work:
<h2>Test if the commands work locally within a Laravel Fly app's project directory</h2>

❓  flyctl deploy --depot
✅ Successfully deploys 
```
admin_kath@__6:~/.../newapp$ /.../go/flyctl/bin/flyctl.exe deploy --depot
==> Verifying app config
Validating /.../newapp/fly.toml
✓ Configuration is valid
--> Verified app config
==> Building image
==> Building image with Depot
--> build:  (​)
[+] Building 47.1s (22/22) FINISHED                 
```
❓  flyctl machines run --build-depot 
✅  Successfully creates a new machine for the app
```
admin_kath@__6:~/.../newapp$ /.../go/flyctl/bin/flyctl.exe  machines run --build-depot .
==> Building image with Depot
--> build:  (​)
[+] Building 17.2s (22/22) FINISHED     

... // OTHER LOGS HERE!

Image size: 125 MB

Success! A Machine has been successfully launched in app newapp-spring-frost-6423
 Machine ID: _____
 Instance ID: ___
 State: created
       
````

<h2> Check if the both commands work as expected s in a CI environment </h2>
- Copied a compressed version of the local build of the PR's flyctl.exe to a repository
- Created a github workflow to use `flyctl.exe deploy --depot` and `flyctl.exe machines run --build-depot` [here's a sample run](https://github.com/KTanAug21/my-sample-app/actions/runs/10166790875/job/28117747227) 
- Command run successfully!

File: internal/build/imgsrc/depot.go [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+func (d *DepotBuilder) Run(ctx context.Context, _ *dockerClientFactory, streams *iostreams.IOStreams, opts ImageOptions, build *build) (*DeploymentImage, string, error) {
+	ctx, span := tracing.GetTracer().Start(ctx, "depot_builder", trace.WithAttributes(opts.ToSpanAttributes()...))
+	defer span.End()
+
+	build.BuildStart()
+
+	var dockerfile string
+
+	if opts.DockerfilePath != "" {
```
Comment (rugwirobaker) [inline internal/build/imgsrc/depot.go:48]: nit: a switch statement is so much easier than of `if-else` clauses

File: internal/build/imgsrc/depot.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+		}
+		// On Windows, convert \ to a slash / as the docker build will
+		// run in a Linux VM at the end.
+		relDockerfile = filepath.ToSlash(p)
+	}
+	span.SetAttributes(attribute.String("relative_dockerfile_path", relDockerfile))
+
+	build.ImageBuildStart()
+
+	image, err := depotBuild(ctx, streams, opts, dockerfile, build)
```
Comment (rugwirobaker) [inline internal/build/imgsrc/depot.go:84]: Is there a reason `depotBuild` can't pull `streams *iostreams.IOStreams` out of the context object on its own? Cause I think it makes the function signature easier to read and the remaining args coherent.

File: internal/build/imgsrc/depot.go [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+		region = "fly-" + region
+	}
+
+	buildInfo, err := apiClient.EnsureDepotRemoteBuilder(ctx, &fly.EnsureDepotRemoteBuilderInput{
+		AppName: &opts.AppName,
+		Region:  &region,
+	})
+	if err != nil {
+		streams.StopProgressIndicator()
+		buildState.BuilderInitFinish()
```
Comment (rugwirobaker) [inline internal/build/imgsrc/depot.go:125]: Could `buildStateInitFinish()` be in a defer statement?

Comment (billyb2) [reply internal/build/imgsrc/depot.go:125]: that's a good idea. i'll move the builder init code into a new function

File: internal/build/imgsrc/depot.go (hunk 2 above)
Comment (rugwirobaker) [reply internal/build/imgsrc/depot.go:84]: Cool, I don't have any strong feelings about it

**Review by rugwirobaker — APPROVED** [review]

**Pull request merged**